### DISCUSSION TITLE
Command line query tools

### DISCUSSION THREAD

**Discussion by edrex** [srid/emanote#178] [Ideas]
I'm very interested in creating a general-purpose command line query tool for mdWiki (what I've been calling markdown directory + wiki extensions, need a better name). Since I'm converging on emanote's particular flavor (and also wanting to nudge emanote closer to other tools, esp wrt to relative path handling) building a small Haskell command line wrapper around emanote's functionality seems like a solid approach. A benefit of this would be more testing/improvement in emanote.

It might even make sense to include a very thin query wrapper in the main emanote program:

```
emanote query "tag:foo OR path:/foo/**"
```

I don't have a particular feature set pinned down, and see this more as a progression of prototypes for exploring what's useful.

- Some more flexible graph-walking operators
- Eventually, wiki-wide refactoring (link rename, directory move, date format conversions, and migrating syntax extensions from other tools). Maybe using Pandoc for an intermediate representation? This will require me to really learn Haskell.

One runtime-tooling-heavy approach I might take for advanced prototypes is to slurp the new graph data export (#161) into a general purpose graph DB like Neo4J.

**Comment by srid** [comment]
CLI would be useful. If you are using Emanote as library, check out https://github.com/srid/emanote/blob/master/src/Emanote/Model/Query.hs

**Reply by edrex** [reply]
In terms of performance, IDT emanote does any caching between runs, so using it for oneshot queries might be too slow. Alternatively, what about adding a query endpoint to the server?

**Reply by srid** [reply]
Emanote when run in live server mode (not generate) *incrementally* builds the model based on filesystem events. It's all cached but only in memory.

This is why #161 provides a http interface, rather than a CLI interface ... because that latter would have to be slow (`O(n)`). In future we may be able to utilize something like SQLite for caching. So the live server can incrementally update the on-disk database, which other tools may now directly query instantly.

**Reply by srid** [reply]
But then ... one could leave the live server running as a daemon, and then write their CLI tools such that it simply talks to the localhost http API ...

**Reply by edrex** [reply]
Exactly. A bit more runtime complication, but the extra speed seems worth it.

I should have said: "*IDT emanote does any caching on disk, so the startup overhead starting a oneshot query process might be too much*, since yeah, the incrementally-updating in-memory cache is great.

**Comment by edrex** [comment]
based on your experiences with Pandoc @srid, does using it as an intermediate representation for MD -> MD transformation seem like a good approach?

**Reply by srid** [reply]
Yea, you can see the AST it uses here (basically `Inline` and `Block`, recursively defined):

https://hackage.haskell.org/package/pandoc-types-1.22/docs/Text-Pandoc-Definition.html#t:Pandoc

[pandoc-link-context](https://github.com/srid/pandoc-link-context/blob/master/src/Text/Pandoc/LinkContext.hs) (used in Emanote, to extract backlink context) is an example of processing this AST ... as in AST->AST. Generating Markdown in a lossless manner from the AST is tricky I think, but this is something I've been wanting to do as well (eg: a CLI program that writes a mood log entry to the table in say MoodJournal/2021-09-21.md).

**Reply by edrex** [reply]
Yeah, I've been thinking also about tooling for writing "materialized views" back to markdown, which would allow pipeline integration with other tools. Inspired by https://github.com/tools-life/taskwiki#viewports

**Reply by edrex** [reply]
So basically, you could have special headings (or code blocks) including some computation description, and then the tooling would find all such in the wiki, eval them, and write the results back out to the markdown. A bit scary but i might include some automatic git checks (assert clean WC, stage the result) to avoid mishaps.

**Reply by edrex** [reply]
idk if this is actually a good idea tho. just want to experiment with it.

**Reply by edrex** [reply]
> pandoc-link-context (used in Emanote, to extract backlink context) is an example of processing this AST ... as in AST->AST.

Actually seeing emanote's usage of pandoc-link-context to decorate the AST with backlinks is what got me thinking about using it that way. But I think it's an old idea.. if only I had some kind of decades-spanning external personal knowledge repository I could consult..

**Reply by srid** [reply]
pandoc-link-context just extracts (`query`) it - but doesn't touch (`walk`) the AST itself. The actual rendering of backlinks is done in Heist templates (see backlinks.tpl and timeline.tpl).

(See [here](https://hackage.haskell.org/package/pandoc-types-1.21/docs/Text-Pandoc-Walk.html) for what "query" and "walk" means).

**Reply by edrex** [reply]
Renaming a document by this approach:
 - Check emanote's graph model for documents that need updates
 - Parse/transform/writetemp each doc with Pandoc
 - Move them into place as atomically as possible
 - (with some git stuff in there if enabled)

**Reply by srid** [reply]
@edrex For renaming, perhaps you might want to comment on https://github.com/mickael-menu/zk/discussions/199#discussioncomment-2751992

I really haven't thought much about an `emanote rename` - but it is possible.

**Comment by srid** [comment]
We now export the data as JSON in generated site / liver-server:

https://github.com/srid/emanote/commit/e15a3c930076a076e21b1186ff3ed1ab5301df72

https://emanote.srid.ca/-/export.json
